src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -3326,204 +3404,295 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, Number
+                                    }
+
+                                    break;
+                                case NumberSuffixFlags.BigInteger:
+                                    result = (BigInteger)Math.Round(doubleValue);
+                                    return true;
                             }
 
-                            return true;
+                            // Invalid NumberSuffixFlags combination, or outside bounds of specified type.
+                            result = null;
+                            return false;
                         }
 
-                        // TryParse on (real) number fails.
+                        // TryParse for real numeric literal failed
                         result = null;
                         return false;
                     }
 
-                    if (hex && !strNum[0].IsHexDigit())
+                    if (format == NumberFormat.Hex)
                     {
-                        if (strNum[0] == '-')
+                        if (!strNum[0].IsHexDigit())
+                        {
+                            if (strNum[0] == '-')
+                            {
+                                multiplier = -multiplier;
+                            }
+
+                            strNum = strNum.Slice(1);
+                        }
+
+                        // If we have a hex literal denoting (u)int64, treat it as such
+                        if (suffix == NumberSuffixFlags.None || suffix == NumberSuffixFlags.Unsigned)
                         {
-                            multiplier = -multiplier;
+                            switch (strNum.Length)
+                            {
+                                case 16:
+                                    suffix |= NumberSuffixFlags.Long;
+                                    break;
+                                case 24:
+                                    suffix |= NumberSuffixFlags.Decimal;
+                                    break;
+                                case 32:
+                                    suffix |= NumberSuffixFlags.BigInteger;
+                                    break;
+                            }
                         }
 
-                        strNum = strNum.Substring(1);
+                        // If the string isn't at a length where we expect a signing bit or it's flagged as unsigned
+                        if (suffix.HasFlag(NumberSuffixFlags.Unsigned) || (strNum.Length & 7) != 0)
+                        {
+                            // Allocate new span
+                            Span<char> tempSpan = new char[strNum.Length + 1];
+
+                            // Insert 0 prefix so BigInt doesn't use the high bit as a sign bit
+                            tempSpan[0] = '0';
+
+                            // Copy original digits into the new span after the 0
+                            strNum.CopyTo(tempSpan.Slice(1));
```

Comment:
Interesting thought indeed! I'll have to check through the code paths here and make sure that doesn't break anything. Also, while it does solve the double-allocation in these instances, it also means that there is additional allocation on *all* numerics, and I am worried it may cause the ScanNumberHelper logic some issues. I will investigate this possibility. 😄

Reply:
@iSazonov this would be predicated on having the `strNum` value be a `Span<char>` and not a `ReadOnlySpan<char>`. Is this an acceptable change? I felt `ReadOnlySpan<char>` to be more appropriate, as it was initially a `string` value.

Reply:
My suggestion is to place one space on first place in StringBuilder. Later we can slice with second positions if we don't need the space.

Reply:
Ah, in that case that should be doable, yes! I'll have to ensure it gets added after any `+` or `-` in the number string... no worries. Great idea, thank you!

Reply:
That ought to do the trick! I restricted it to being added only to hex literals, and it seems to work well. 😄